[wzhua02] iP

wzhua02 opened the pull request:
### Baymax
> Your **personal _healthcare_ companion** ~~(not really)~~
- Keeps track of stuff
     - And more.
---
1. Link to `Google` [here](https://www.google.com)
2. Or use other search engines
3. Wow emojis :smiley_cat:

Make sure to:
- [ ] Do your iP
- [ ] Meet with your team for tP
- [x] Do your weekly quiz

```
public void sayGdWork() {
    System.out.println("Good work");
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wzhua02 committed:
Renamed Duke to Baymax

wzhua02 committed:
Did Level 1: Echo and Level 2: Add, List

wzhua02 committed:
Converted tasks from just Strings to the class Task; input is now split to String[], taskList uses ArrayList to store

wzhua02 committed:
Done level 3; Change switch-case to simply if-else

wzhua02 committed:
Done Level-4

wzhua02 committed:
Added new classes for each Task as files and with inheritance; Changed back to switch-case.

wzhua02 committed:
Done Level 5; Added Exception class

wzhua02 committed:
Done A-TextUiTesting

wzhua02 committed:
Done Level-6; Modified TextUiTesting to test it

wzhua02 committed:
Modify reply() method

Removed the indent string from the method, implemented a method
that auto indents every System.out message.

wzhua02 committed:
Add tasks.txt

wzhua02 committed:
Level-7

Refactor tasks.txt.

Add loading and saving to tasks.txt to Baymax.

Add saveFormat method and overload constructor to Task and its
subclasses to support saving and loading feature.

wzhua02 committed:
Merge branch 'branch-Level-7'

wzhua02 committed:
Remove code that was used to test some
functionality of the chatbot.

wzhua02 committed:
Add DateTime parsing to Baymax

Modify Deadline and Event to parse and output DateTime

wzhua02 committed:
Modifications for text-ui-test

Fix text bug in Event.java.

Remove trailing whitespaces in messages in Baymax.java.

Update runtest.bat, input.txt and EXPECTED.TXT for level 8.

wzhua02 committed:
Merge branch 'branch-Level-8'

wzhua02 committed:
Refactor Baymax for improved OOP design

The functionality for file interaction, task management, and user
interaction is tightly coupled in `Baymax.java`. Tasks have a redundant
type parameter, and the `Task` class has insufficient visibility scope.

This tight coupling makes the codebase harder to maintain and extend.
The redundant type parameter adds unnecessary complexity without
providing any benefits.

Let's refactor the code to adhere to object-oriented principles by
introducing dedicated classes for specific responsibilities:
- Add a `Storage` class to encapsulate file interaction functionality.
- Add a `TaskList` class to manage the task list and its interactions.
- Add a `UserInterface` class to handle user input and Baymax's
responses.
- Refactor `Baymax.java` to align with this modular design, leaving
parser functionality to be extracted separately.
- Remove the type parameter from all `Task` subclasses to reduce
complexity and redundancy.
- Make the `Task` class public to enforce proper scope and visibility.

This approach aligns with the Single Responsibility Principle (SRP) and
improves code readability, modularity, and testability. Tasks no longer
need a redundant type parameter, as constructors already imply the type.

Further parser refactoring will be addressed in a future commit.

wzhua02 committed:
Add parser class

Extract command logic functionality to Parser class.

wzhua02 committed:
Merge branch 'branch-A-MoreOOP'

wzhua02 committed:
Modify functionality of Parser.java

Moved while loop to Baymax.java.

Moved parseDateTime function to Parser.java.

Renamed UserInterface to Ui.

wzhua02 committed:
Merge pull request #2 from 'branch-A-MoreOOP'

wzhua02 committed:
Refactor all class files into respective packages

Modify runtest.bat to detect the Baymax.java in new packages.

Modify .gitignore to account for refactored files.

wzhua02 committed:
Merge request #3 branch 'branch-A-MoreOOP'

wzhua02 committed:
Merge branch 'add-gradle-support'

wzhua02 committed:
Fix a bug in Parser.java

Redundant scan was found in finally statement

wzhua02 committed:
Gradle is able to run Baymax

wzhua02 committed:
Add JUnit Tests for Storage, Ui and Parser

wzhua02 committed:
Update gradle build file to name jar file

wzhua02 committed:
Bug fix: Gradle build file name fix

wzhua02 committed:
Add javadocs to all classes

wzhua02 committed:
Modify files to meet coding standard, excluding JavaDoc comments

wzhua02 committed:
Modify Parser and TaskList

Abstracted create different tasks functionality from TaskList to Parser,
to follow single responsibility rule for TaskList methods.

Added a new case "find", functionality will be implemented in future
commit

wzhua02 committed:
Add find command functionality

Add getTaskIdx method to TaskList

Add getDescription method to Task

Add a "find" case in Parser.java

wzhua02 committed:
Merge request #1 from branch 'branch-A-JavaDoc'

wzhua02 committed:
Merge request #2 from branch 'branch-A-CodingStandard'

wzhua02 committed:
Merge request #1 from branch 'branch-Level-9'

wzhua02 committed:
Update javadocs for Task and TaskList

wzhua02 committed:
Changes to some methods and code quality fixes

Code quality fixed in BaymaxException and Storage.

Partially changed Ui to static context. Further changes to make in future commits.

Modified list tasks functionality.

wzhua02 committed:
Modify Ui to be fully static

Modified UiTest to comply with static context.

wzhua02 committed:
Add javadocs and code quality fixes

clx3210 commented on src/main/java/baymax/Baymax.java:
```diff
@@ -0,0 +1,51 @@
+    private final Ui ui;
+
+    /**
+     * Constructs a Baymax instance.
```
Shall we add a newline here between the main description and the `@param`?

clx3210 commented on src/main/java/baymax/Baymax.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * The main method to start the chatbot.
```
Same comment as above.

clx3210 commented on src/main/java/baymax/io/Storage.java:
```diff
@@ -0,0 +1,78 @@
+     * @return an ArrayList containing the tasks read from the file
+     */
+    public ArrayList<String> load() {
+        ArrayList<String> returnList = new ArrayList<>();
```
Shall we rename this to a more meaningful noun? For example, how about `rawDataObjects`?

clx3210 commented on src/main/java/baymax/io/Ui.java:
```diff
@@ -0,0 +1,78 @@
+ */
+public class Ui {
+    private static final String INDENT = "    ";
+    private final Scanner scan;
```
Shall we rename to the `scanner` noun instead?

clx3210 commented on src/main/java/baymax/task/Deadline.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Represents a deadline task with a specific due date and time.
+ */
+public class Deadline extends Task {
```
I like the naming convention used for this class, It is very clear.

clx3210 commented on src/main/java/baymax/task/Task.java:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @param mark {@code true} to mark as done, {@code false} to unmark.
+     */
+    public void marker(boolean mark) {
```
Shall we rename `marker` to a verb? For example:
```java
public void markDoneStatus(boolean isDone) {
...
}
```

wzhua02 replied:
good point, have changed the method to just be markTask() and unmarkTask() instead.

clx3210 commented on src/main/java/baymax/task/TaskList.java:
```diff
@@ -0,0 +1,147 @@
+     *
+     * @return The size of the task list.
+     */
+    public int size() {
```
How about renaming this to `getSize`?

wzhua02 replied:
good point, will change this.

clx3210 commented on src/main/java/baymax/util/Parser.java:
```diff
@@ -0,0 +1,196 @@
+                if (idx < 0 || idx >= tasks.size()) {
+                    throw new BaymaxException("I do not know which task you are referring to.");
+                }
+                Task theTask = tasks.getTask(idx);
```
How about `taskToUnmark`? This may help improve readability.

wzhua02 replied:
Yup will be adding a method unmarkTask in the next few commits

clx3210 commented on src/main/java/baymax/util/Parser.java:
```diff
@@ -0,0 +1,196 @@
+                if (idx < 0 || idx >= tasks.size()) {
+                    throw new BaymaxException("I do not know which task you are referring to.");
+                }
+                Task theTask = tasks.getTask(idx);
```
Same comment as above

clx3210 commented on src/main/java/baymax/util/Parser.java:
```diff
@@ -0,0 +1,196 @@
+                if (args.length < 2) {
+                    throw new BaymaxException("Let me know what task you would like to find.");
+                }
+                ArrayList<Task> foundTasks = tasks.findTask(args[1]);
```
I like the variable naming here! The variable naming here provides a clear idea of what the code is doing.

clx3210 commented on src/test/java/baymax/io/StorageTest.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Test
+    void testSaveAndLoad() {
```
Shall we rename the methods here to use the naming convention for testing methods? Example: `saveAndLoad_multipleTasks_success()`

clx3210 commented on src/test/java/baymax/io/UiTest.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @Test
+    void testGetInput() {
```
Same comment as above for all the unit test methods

clx3210 reviewed (commented):
Looks good! 👍 
Just some minor comments on code quality naming guidelines.

wzhua02 committed:
Added GUI to Baymax

Remove Ui class, with GUI there is no use cases for methods in Ui class.

wzhua02 committed:
Add and modify javadocs in Baymax and Parser

wzhua02 committed:
Refactor Storage from io package to util package

wzhua02 committed:
Modify ParserTest and StorageTest

wzhua02 committed:
Modify image resource and minor text bug fix

wzhua02 committed:
Add bye command functionality

wzhua02 committed:
Dialog box changes color for certain commands

wzhua02 committed:
Add background to app window

wzhua02 committed:
Merge branch 'branch-Level-10'

zihung20 commented on src/main/java/baymax/task/TaskList.java:
```diff
@@ -0,0 +1,168 @@
+ * Manages a list of tasks, allowing addition, removal, retrieval, and persistence.
+ */
+public class TaskList {
+    private final ArrayList<Task> taskList;
```
Shouldn't this to be named as plural?`private final ArrayList<Task> taskLists `

iuhiah replied:
I think `taskList` would be the better option! This class would only represent one task list that contains multiple tasks, so it would either be `taskList` or `tasks`!

zihung20 commented on src/main/java/baymax/task/Deadline.java:
```diff
@@ -0,0 +1,71 @@
+package baymax.task;
```
I like that you put all related classes into a package

zihung20 commented on src/main/java/baymax/util/Parser.java:
```diff
@@ -0,0 +1,174 @@
+     * @param input   The user input string.
+     * @param tasks   The TaskList containing user tasks.
+     * @param storage The Storage handler for saving and loading tasks.
+     * @return A String that Baymax will reply to the user based on the input.
```
Isn't this should be An array of String?

wzhua02 replied:
True good catch! Planning to change this into command class anyway, will be abstracting command functionality into individual classes.

zihung20 reviewed (commented):
Overall, you strictly followed the coding standard, and there aren't many coding standard issues in your files. Good job keeping the code standardized.

iuhiah commented on src/main/java/baymax/task/Event.java:
```diff
@@ -0,0 +1,87 @@
+     *
+     * @return A formatted string representing the event task.
+     */
+    @Override
+    public String toString() {
+        String typeStr = "[" + this.type + "]";
+        String markStr = "[" + getStatusIcon() + "] ";
+        String fromToStr = " (from: " + getFromDate() + ") (to: " + getToDate() + ")";
+        return typeStr + markStr + getDescription() + fromToStr;
+    }
+
+    /**
+     * Converts the event task into a formatted string for saving.
+     *
+     * @return A formatted string representing the event for storage.
+     */
+    @Override
+    public String toSaveFormat() {
+        return "E | " + (isDone ? "1" : "0") + " | " + getDescription()
+                + " | " + this.fromDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"))
+                + " | " + this.toDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
+    }
```
```suggestion
    @Override
    public String toString() {
        String typeStr = "[" + this.type + "]";
        String fromToStr = " (from: " + getFromDate() + ") (to: " + getToDate() + ")";
        return typeStr + super.toString() + fromToStr;
    }

    /**
     * Converts the event task into a formatted string for saving.
     *
     * @return A formatted string representing the event for storage.
     */
    @Override
    public String toSaveFormat() {
        return "E | " + super.toSaveFormat()
                + " | " + this.fromDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"))
                + " | " + this.toDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
    }
```

Perhaps move the common substrings to the superclass for better abstraction and code management!

wzhua02 replied:
Good suggestion, I think that looks much cleaner, thanks!

iuhiah commented on src/main/java/baymax/util/Parser.java:
```diff
@@ -0,0 +1,174 @@
+                if (idx < 0 || idx >= tasks.size()) {
+                    throw new BaymaxException("I do not know which task you are referring to.");
+                }
+                Task theTask = tasks.getTask(idx);
+                theTask.marker(true);
+                tasks.save(storage);
```
Not a coding quality standard comment(?), but I would think about only asking the Parser class to parse data and nothing more (e.g. adding the task, saving tasks to storage)! This will help in isolating the class for testing in case there are other issues with the TaskList and Storage classes :)

wzhua02 replied:
Yup have been planning to create classes for each command to separate their functionality from this method, will be doing this in the next few commits, but thanks for pointing it out.

iuhiah commented on src/main/java/baymax/util/Parser.java:
```diff
@@ -0,0 +1,174 @@
+     * @throws BaymaxException If the date-time string does not match any valid format.
+     */
+    public static LocalDateTime parseDateTime(String dateTimeStr) throws BaymaxException {
+        String[] patterns = {
+            "yyyy-MM-dd HH:mm", //e.g. 2025-01-27 12:30
+            "dd/MM/yyyy HH:mm", //e.g. 27/01/2025 12:30
+            "yyyy MM dd HH:mm", //e.g. 2025 01 27 12:30
+        };
```
Consider setting this as a static class variable! Else a new DateTimeFormatter object will be created whenever this method is called.

wzhua02 replied:
Good point!

iuhiah reviewed (commented):
Project is well organised into packages! Some nits have already been pointed out by other reviewers (e,g, getter naming convention). I would think about dereferencing class if possible (as mentioned in the Parser class) as it seems quite tough to carry out isolated unit tests with the current structure.

P.S. Baymax concept is super cute!

wzhua02 committed:
Refactor command functionality into classes
Command functionality is conducted in Parser class.

Difficult to test Parser class, the class should only parse data and
not anything else. Single functionality rule.

Let's create a command class for each command and isolate each command's
functionality in each class.

This makes the project's structure clearer and separates the command
functionality from the Parser class.

wzhua02 committed:
Add dialog styles for each command type

wzhua02 committed:
Update javadocs for all classes

wzhua02 committed:
Abstract common substrings of task classes

wzhua02 committed:
Rename size to getSize in TaskList

wzhua02 committed:
Change patterns to static class variable in Parser class

wzhua02 committed:
Implement JUnit Test for Parser and Storage

wzhua02 committed:
Add some exceptions in Parser

wzhua02 committed:
Modify files for code quality

wzhua02 committed:
Merge branch 'A-EvenMoreOOP'

wzhua02 committed:
Add assertions and minor code quality fixes

wzhua02 committed:
Add streams

wzhua02 committed:
Fix minor code quality issue

wzhua02 committed:
Code quality fixes

Parser.parse method has many if statements that throws a BaymaxException
with a unique message.

Refactored the method to improve readability.

Extracting the if and throw behaviour into a separate method, and the
statement in each switch case is more readable.

Let's,
* create a method to check if there are arguments and throw an exception
if there are none with a unique message for each command.
* create a method to get the index of the desired string that indicates
the start of a parameter for the command and throw an exception if it
cannot be found with a unique message for each command.

wzhua02 committed:
Merge branch 'branch-A-Assertions'

wzhua02 committed:
Merge branch 'master' into branch-A-CodeQuality

wzhua02 committed:
Merge branch 'master' into branch-A-Streams

wzhua02 committed:
Merge branch 'branch-A-CodeQuality'

wzhua02 committed:
Merge branch 'master' into branch-A-Streams

wzhua02 committed:
Merge branch 'branch-A-Streams'

wzhua02 committed:
Add CI

wzhua02 committed:
Remove text-ui-tests in gradle.yml

wzhua02 committed:
Make Task Comparable

wzhua02 committed:
Add Sort Command and sort functionality

wzhua02 committed:
Add parse for sort

wzhua02 committed:
Fix javadocs for SortCommand class

wzhua02 committed:
Merge branch 'branch-C-Sort'

wzhua02 committed:
Remove StorageTest due to bugs

wzhua02 committed:
Fix minor code quality

wzhua02 committed:
Remove throw exception from getResonse

wzhua02 committed:
Merge pull request #2 from wzhua02/branch-GitMerge

Remove throw exception from getResonse

wzhua02 committed:
Add credit to reused code

wzhua02 committed:
Modify header comments in GUI for code quality

wzhua02 committed:
Edit credit comment in TaskList

wzhua02 committed:
Modify parse method to SLAP better

wzhua02 committed:
Edit Readme.md

wzhua02 committed:
Fix initialization of Baymax

wzhua02 committed:
Modify README of project

wzhua02 committed:
Modify Ui screenshot

wzhua02 committed:
Modify MainWindow initialize

wzhua02 committed:
Modify build.gradle to build v0.2

wzhua02 committed:
Update README.md

wzhua02 committed:
Merge branch 'master' of https://github.com/wzhua02/ip

wzhua02 committed:
Add minor fix to FILE_PATH

wzhua02 committed:
Fix bugs in error handling

wzhua02 committed:
Modify DateTime parsing to follow only one pattern

wzhua02 committed:
Fix minor comment grammer error

wzhua02 committed:
Update README.md

wzhua02 committed:
Update README.md

wzhua02 committed:
Fix bug in adding event and deadline tasks

wzhua02 committed:
Fix command input bug

wzhua02 committed:
Fix bug to process integers and fix for AddTodo
